File: balanceBinaryTree.java
// Time Complexity :O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes in understanding the balanced binary tree

class Solution {
    public boolean isBalanced(TreeNode root) {
       if(helper(root) != -1) return true;
        return false;
    }
    private int helper(TreeNode root){
        if(root == null) return 0;
        int left = helper(root.left);
        if(left == -1) return -1;
        int right = helper(root.right);
        if(left == -1 || right == -1) return -1;
        int diff = Math.abs(left - right);
        if(diff > 1) return - 1;
        return Math.max(left, right)+ 1;
    }
}

File: BalancebinaryTree.py
# Time Complexity : O(n) for each operation.
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

# Approach is to find the height of left and right subtree and recursively check for every node.

class Solution:
    def heightofBT(self, root):
        if(root==None):
            return 0
        lh=self.heightofBT(root.left)
        rh=self.heightofBT(root.right)
        
        if(lh>rh):
            return lh+1
        else:
            return rh+1
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if(root==None):
            return True
        
        lh=self.heightofBT(root.left)
        
        rh=self.heightofBT(root.right)
        left_bal=self.isBalanced(root.left)
        right_bal=self.isBalanced(root.right)
        height_diff = abs(lh - rh)
        return height_diff <= 1 and left_bal and right_bal